### PR TITLE
try to rename chef_install_dir if files still exist after nuking it

### PR DESCRIPTION
**Pull request by Happycoil**
### Description

This seeks to address problems upgrading chef-client due to open handles from other processes on files in the chef install dir.

We may want to be able to toggle this on/off.

### Issues Resolved

https://github.com/chef-cookbooks/chef_client_updater/issues/110

### Check List

- [ ] All tests pass. See <https://github.com/chef-cookbooks/community_cookbook_documentation/blob/master/TESTING.MD>
- [ ] New functionality includes testing.
- [ ] New functionality has been documented in the README if applicable
- [ ] All commits have been signed for the Developer Certificate of Origin. See <https://github.com/chef-cookbooks/community_cookbook_documentation/blob/master/CONTRIBUTING.MD>

### CODE REVIEW

**Comment by Happycoil** [comment]
Fixed the missing DCO

**Comment by lamont-granquist** [comment]
backstory is in #67 and #69

**Comment by tas50** [comment]
@Happycoil Can you sign this off for the DCO so we can merge / release it?

**Comment by Happycoil** [comment]
@tas50 I believe I already did.

I take it we're not adding a parameter to toggle this workaround? I feel like it should at least emit a warning, but that's kinda hard with the script running in a separate task and all...

How about I take a stab at adding logging to the script in a another PR?

**Comment by kitforbes** [comment]
@Happycoil it looks like you manually typed the sign off, and you are missing the chevrons around your email address which could be causing the failure here.

```
Signed-off-by: Tor Magnus Rakvåg tm@intility.no
Signed-off-by: Tor Magnus Rakvåg <tm@intility.no>
```

I've been testing this change and it works great. I can see some more improvements to this PowerShell script that I will submit once this has been merged.

Logging actions from this script to `chef_upgrade.log` in the same directory as `chef_upgrade.ps1` would be a great addition. I added a basic version when I was trying to debug some issues myself. I updated the code for the scheduled task to be the below. Basic but effective.

```ruby
def prepare_windows
  copy_opt_chef(chef_install_dir, chef_backup_dir)
  Kernel.spawn("c:/windows/system32/schtasks.exe /F /RU SYSTEM /create /sc once /ST \"#{upgrade_start_time}\" /tn Chef_upgrade /tr \"powershell.exe -ExecutionPolicy Bypass -Command '& c:/opscode/chef_upgrade.ps1 *>c:\\opscode\\chef_upgrade.log'\"")
  FileUtils.rm_rf "#{chef_install_dir}/bin/chef-client.bat"
end
```

**Comment by Happycoil** [comment]
There we go, it was copy pasted from a previous commit where I forgot the DCO...

Yeah, simply sending all the streams to a file is probably the most effective logging solution.

**Comment by Happycoil** [comment]
@kitforbes Maybe we should reorder the script to rename and then delete. I haven't encountered that scenario though, can you repro it reliably?

**Comment by kitforbes** [comment]
If we rename first, then there will be no need to remove the install directory as it will no longer exist. I have uploaded my working changes to [my fork](https://github.com/kitforbes/chef_client_updater/blob/fix-update-script/providers/default.rb) if you are curious about my changes.

I can't create a reliable repro as the permission issue I've been encountering only sometimes occurs during the recursive remove. There is something in my environment which is causing the DHCP Client to lock the `chef-log.dll`, preventing the deletion. My approach is working on my test machines, but I haven't tried this on a real-world server yet.

**Comment by lamont-granquist** [comment]
@smurawski are you still around?

@Happycoil renaming is what we used to do (via FileUtils.mv though) and then there was the big discussion which originally started off with renaming via powershell and which settled on this implementation.

**Comment by RobbertJanSW** [comment]
I've tried renaming earlier as well while implementing the current way Windows upgrades. I have to be honest I dont remember whether I had more issues I ran into, but at least one issue was this:

Windows is able to keep the lock on that specific dll while renaming the folder above it. So the upgrade task resulting from this commit will still fail (at least in some cases) during the attempt of removing the locked_files folder. 

@lamont-granquist there's a huge difference between the FileUtils.mv and Powershell rename-item.
FileUtils.mv does iterate over all files one by one, moving them one by one. The older implementation did a forced move, which in all cases succeeded (it rescued all errors it ran in to, but did stop the move process after running into locked files, returning succes and two folders with half the content). Whereas the powershell-rename will either end up with everything in the new folder, of everything in the old. No half-way copy.

**Comment by Happycoil** [comment]
Unless we want to (try to) kill the external processes with locks somewhere in the folder I don't think there is a sure-fire way of removing the chef install dir. This way lets us do the upgrade even if there are open handles from external processes, but at the cost of leaving a weird folder.

Pretty sure we have to remove the locked files folder before any subsequent upgrades are attempted though...

**Comment by kitforbes** [comment]
I don't know enough about Windows to say either way. I'll test whatever you settle on to see if it gets past my current issue.

**Comment by Happycoil** [comment]
I added a commit which cleans up the locked files dir, but I can't decide whether it should raise or just warn like it does now.

The folder being being present when the upgrade script runs will case it to fail, and we don't have any logging of that script yet. This makes me lean in the direction of raising, but that will almost always case the run immediately after upgrading to fail, since the lock is probably still there.

If we leave it like this we just keep warning and converging while we hope that the locks are gone before we try upgrading again. We can of course make the upgrade script rename any remnant folders from previous upgrades...

Thoughts?

**Comment by kitforbes** [comment]
Warning instead of raising seems reasonable to me. Cleaning up the directory on future runs - as it currently does with the backup directory - makes sense. From my earlier testing I found that the lock would eventually release, and in this situation I wouldn't mind if that was 2+ runs later.

It is an unlikely scenario that someone would upgrade so soon after the previous upgrade, but it could certainly happen. Handling that edge case would make it more robust.

**Comment by Vasu1105** [comment]
I tried to test with your changes to upgrade from chef 12.19.36 to 13.8.5 but upgrade still fails for me with 

```

**Comment by Vasu1105** [comment]
I have verified this changes by updating chef from 12.19.36 to 13.8.5 and updating 13.8.5 from 14.0.190.
Version are updated successfully verified with chef-client -v command.
Also verified that scheduled task is getting created chef.upgrade is getting created. I haven't come across any chef.locked_files though. This has been tried on Windows 2012R2.

File: providers/default.rb [hunk 1]
```diff
@@ -191,13 +191,26 @@ def chef_backup_dir
...
     converge_by("remove #{chef_backup_dir} from previous chef-client run") do
       FileUtils.rm_rf chef_backup_dir
     end
   end
+  if ::File.exist?(chef_locked_files_dir) # rubocop:disable Style/GuardClause
+    converge_by("remove #{chef_locked_files_dir} from previous chef-client run") do
+      begin
+        FileUtils.rm_r chef_locked_files_dir
+      rescue
+        Chef::Log.warn("Failed to remove locked files dir!")
```
Comment (btm) [inline providers/default.rb]: `"remove #{chef_locked_files_dir} from previous chef-client run"` this is more accurately a previous upgrade, right? should we say `"remove #{chef_locked_files_dir} from previous chef client upgrade"`?

File: providers/default.rb [hunk 2]
```diff
@@ -191,13 +191,26 @@ def chef_backup_dir
...
 def cleanup
   if ::File.exist?(chef_backup_dir) # rubocop:disable Style/GuardClause
     converge_by("remove #{chef_backup_dir} from previous chef-client run") do
       FileUtils.rm_rf chef_backup_dir
     end
   end
+  if ::File.exist?(chef_locked_files_dir) # rubocop:disable Style/GuardClause
+    converge_by("remove #{chef_locked_files_dir} from previous chef-client run") do
+      begin
+        FileUtils.rm_r chef_locked_files_dir
```
Comment (btm) [inline providers/default.rb]: Why not `FileUtils.rm_rf`?

File: providers/default.rb [hunk 3]
```diff
@@ -191,13 +191,26 @@ def chef_backup_dir
...
+end
+
 # cleanup cruft from *prior* runs
 def cleanup
   if ::File.exist?(chef_backup_dir) # rubocop:disable Style/GuardClause
     converge_by("remove #{chef_backup_dir} from previous chef-client run") do
       FileUtils.rm_rf chef_backup_dir
     end
   end
+  if ::File.exist?(chef_locked_files_dir) # rubocop:disable Style/GuardClause
```
Comment (btm) [inline providers/default.rb]:
```
providers/default.rb:200:37: W: Lint/UnneededDisable: Unnecessary disabling of Style/GuardClause.
  if ::File.exist?(chef_backup_dir) # rubocop:disable Style/GuardClause
                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

File: providers/default.rb [hunk 4]
```diff
@@ -191,13 +191,26 @@ def chef_backup_dir
...
+    converge_by("remove #{chef_locked_files_dir} from previous chef-client run") do
```
Comment (btm) [inline providers/default.rb]: `"remove #{chef_locked_files_dir} from previous chef-client run"` this is more accurately a previous upgrade, right? should we say `"remove #{chef_locked_files_dir} from previous chef client upgrade"`?

File: providers/default.rb (hunk 1 above)
Comment (btm) [inline providers/default.rb]: I suggest putting the path in here so if you see the warning you know immediately what path to go look at. If you don't put the path in, you should switch to single-quoted strings.
```
Chef::Log.warn("Failed to remove locked files directory #{chef_locked_files_dir}")
```
```
providers/default.rb:210:24: C: Style/StringLiterals: Prefer single-quoted strings when you don't need string interpolation or special symbols.
        Chef::Log.warn("Failed to remove locked files dir!")
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

File: providers/default.rb (hunk 3 above)
Comment (Happycoil) [reply providers/default.rb]: Fixed

File: providers/default.rb (hunk 4 above)
Comment (Happycoil) [reply providers/default.rb]: Makes sense, fixed

File: providers/default.rb (hunk 1 above)
Comment (Happycoil) [reply providers/default.rb]: Fixed (do we use that cop?)

**Comment by thommay** [comment]
Superseded by #117
